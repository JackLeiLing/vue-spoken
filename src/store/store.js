import Vue from "vue";
import Vuex from "vuex"
import teachersAPI from 'api-client/teachers'
import _set from 'lodash/set'
import createPersistedState from "vuex-persistedstate";

Vue.use(Vuex);

export default new Vuex.Store({
  plugins: [createPersistedState()],
  state: {
    teachers: []
  },
  mutations: {
    setTeachers(state, payload) {
      state.teachers = payload
    },
    updateTeachers(state, { id, objKey, value }) {
      let teacher = state.teachers.find(t=>t.id===id)
      console.log(teacher)
      _set(teacher,objKey,value)
      state.teachers=[...state.teachers]
    }
  },
  actions: {
    getTeachers({ commit }) {
      teachersAPI.getTeachers().then((r) => {
        commit('setTeachers', r)
      }).catch((e)=>{
        console.log('error', e)
      })
    }
  },
  getters: {
    getTeacherByID: (state) => (id) => {
      return state.teachers.find(teacher => teacher.id === id)
    }
  }
});
