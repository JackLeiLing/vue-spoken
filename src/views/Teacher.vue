<template>
  <v-container class="editable-form white--text">
    <!-- <v-row>
      <v-col  :cols="cols(v)" :key="k" v-for="(v, k) in teacher">
        <Field
          :id="teacher.id"
          :objKey="k"
          :value="v"
          @fieldChange="updateTeacher"
        />
      </v-col>
    </v-row> -->
    <v-row>
      <v-col v-for="field in teacherForm" :key="field.apiKey">
         <Field
          :id="teacherID"
          :objKey="field.apiKey"
          :label="field.label"
          :value="getFieldValue(teacher, field.apiKey)"
          @fieldChange="updateTeacher"
          :keyToLabel="field.keyToLabel"
          :validation="field.validation"
        />
      </v-col>
    </v-row>

    <div :key="k + 1" v-for="(v, k) in teacher">{{ k }}:{{ v }}</div>
  </v-container>
</template>

<script>
// @ is an alias to /src
import Field from "@/components/Field";
import teacherDef from "@/api/data/teacherDef";
import _merge from "lodash/merge";
import _get from "lodash/get";

export default {
  name: "TeacherView",
  components: { Field },
  computed: {
    teacherID(){
      return  this.$route.params.id;
    },
    teacher() {
      let id = this.$route.params.id;
      return _merge(teacherDef,this.$store.getters.getTeacherByID(id));
    },
    teacherForm(){
      return [
        {
          apiKey:'firstName',
          label:'First name',
          validation(v){
            if(v.length>10){
              return 'Name length can not be longer than 10'
            }
          }
        },
         {
          apiKey:'lastName',
          label:'Last name'
        },
         {
          apiKey:'students',
          label:'Students'
        },
        {
          apiKey:'address',
          label:'Address',
          keyToLabel:{
            street:'Street',
            state:'State'
          }
        }
      ]
    }
  },
  methods: {
    updateTeacher({ value, objKey, id }) {
      this.$store.commit("updateTeachers", {
        id,
        objKey,
        value
      });
    },
    cols(v){
      if(typeof v === 'object'|| Array.isArray(v)){
        return '12'
      }
      return '3'
    },
    getFieldValue(obj, key){
      return _get(obj, key)
    }

  }
};
</script>
