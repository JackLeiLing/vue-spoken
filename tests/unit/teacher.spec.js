import { mount, createLocalVue} from "@vue/test-utils";
import Teacher from "@/views/teacher.vue";
import VueRouter from 'vue-router'
import store from '../../src/store/store'
import Vuetify from 'vuetify'
import Vuex from 'vuex'
import Vue from 'vue'

// See: https://github.com/vuetifyjs/vuetify/issues/4964.
// See: https://github.com/vuetifyjs/vuetify/issues/4068.
Vue.use(Vuetify)

const localVue = createLocalVue()
localVue.use(VueRouter)
const router = new VueRouter(
  {
    routes: [
      { path: '/teacher/:id', name: 'teacher' }
    ]
  }
)

localVue.use(Vuetify)
localVue.use(Vuex)

describe("teacher.vue", () => {
  it("renders teacher form", async () => {
    await store.dispatch('getTeachers')
    const wrapper = mount(Teacher, {
      localVue,
      router,
      store,
    });
    await router.push({name:'teacher', params:{id:'a'}})
    expect(wrapper.html()).toMatchSnapshot();
  });
});
