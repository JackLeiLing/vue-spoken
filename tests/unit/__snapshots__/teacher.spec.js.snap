// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`teacher.vue renders teacher form 1`] = `
<div class="container editable-form white--text">
  <div class="row">
    <div class="col">
      <div>
        <div class="v-input white--text v-input--is-label-active v-input--is-dirty theme--dark v-text-field">
          <div class="v-input__control">
            <div class="v-input__slot">
              <div class="v-text-field__slot"><label for="input-5" class="v-label v-label--active theme--dark" style="right: 0px; position: absolute;">First name</label><input id="input-5" type="text"></div>
            </div>
            <div class="v-text-field__details">
              <div class="v-messages theme--dark"><span name="message-transition" tag="div" class="v-messages__wrapper"></span></div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div class="col">
      <div>
        <div class="v-input white--text v-input--is-label-active v-input--is-dirty theme--dark v-text-field">
          <div class="v-input__control">
            <div class="v-input__slot">
              <div class="v-text-field__slot"><label for="input-9" class="v-label v-label--active theme--dark" style="right: 0px; position: absolute;">Last name</label><input id="input-9" type="text"></div>
            </div>
            <div class="v-text-field__details">
              <div class="v-messages theme--dark"><span name="message-transition" tag="div" class="v-messages__wrapper"></span></div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div class="col">
      <div>
        <div class="field-collection">
          <h3>Students</h3>
          <div>
            <div class="v-input white--text v-input--is-label-active v-input--is-dirty theme--dark v-text-field">
              <div class="v-input__control">
                <div class="v-input__slot">
                  <div class="v-text-field__slot"><label for="input-31" class="v-label v-label--active theme--dark" style="right: 0px; position: absolute;">Students-0</label><input id="input-31" type="text"></div>
                </div>
                <div class="v-text-field__details">
                  <div class="v-messages theme--dark"><span name="message-transition" tag="div" class="v-messages__wrapper"></span></div>
                </div>
              </div>
            </div>
          </div>
          <div>
            <div class="v-input white--text v-input--is-label-active v-input--is-dirty theme--dark v-text-field">
              <div class="v-input__control">
                <div class="v-input__slot">
                  <div class="v-text-field__slot"><label for="input-35" class="v-label v-label--active theme--dark" style="right: 0px; position: absolute;">Students-1</label><input id="input-35" type="text"></div>
                </div>
                <div class="v-text-field__details">
                  <div class="v-messages theme--dark"><span name="message-transition" tag="div" class="v-messages__wrapper"></span></div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div class="col">
      <div>
        <div class="field-group">
          <h3>Address</h3>
          <div>
            <div class="v-input white--text v-input--is-label-active v-input--is-dirty theme--dark v-text-field">
              <div class="v-input__control">
                <div class="v-input__slot">
                  <div class="v-text-field__slot"><label for="input-15" class="v-label v-label--active theme--dark" style="right: 0px; position: absolute;">street</label><input id="input-15" type="text"></div>
                </div>
                <div class="v-text-field__details">
                  <div class="v-messages theme--dark"><span name="message-transition" tag="div" class="v-messages__wrapper"></span></div>
                </div>
              </div>
            </div>
          </div>
          <div>
            <div class="v-input white--text v-input--is-label-active v-input--is-dirty theme--dark v-text-field">
              <div class="v-input__control">
                <div class="v-input__slot">
                  <div class="v-text-field__slot"><label for="input-19" class="v-label v-label--active theme--dark" style="right: 0px; position: absolute;">suburb</label><input id="input-19" type="text"></div>
                </div>
                <div class="v-text-field__details">
                  <div class="v-messages theme--dark"><span name="message-transition" tag="div" class="v-messages__wrapper"></span></div>
                </div>
              </div>
            </div>
          </div>
          <div>
            <div class="v-input white--text v-input--is-label-active v-input--is-dirty theme--dark v-text-field">
              <div class="v-input__control">
                <div class="v-input__slot">
                  <div class="v-text-field__slot"><label for="input-23" class="v-label v-label--active theme--dark" style="right: 0px; position: absolute;">state</label><input id="input-23" type="text"></div>
                </div>
                <div class="v-text-field__details">
                  <div class="v-messages theme--dark"><span name="message-transition" tag="div" class="v-messages__wrapper"></span></div>
                </div>
              </div>
            </div>
          </div>
          <div>
            <div class="v-input white--text v-input--is-label-active v-input--is-dirty theme--dark v-text-field">
              <div class="v-input__control">
                <div class="v-input__slot">
                  <div class="v-text-field__slot"><label for="input-27" class="v-label v-label--active theme--dark" style="right: 0px; position: absolute;">postCode</label><input id="input-27" type="text"></div>
                </div>
                <div class="v-text-field__details">
                  <div class="v-messages theme--dark"><span name="message-transition" tag="div" class="v-messages__wrapper"></span></div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div>id:a</div>
  <div>firstName:Lei</div>
  <div>lastName:Ling</div>
  <div>nativeLanguage:Mandarin</div>
  <div>image:https://bbts1.azureedge.net/images/p/full/2020/07/4dbb5cf7-f185-47ff-8467-c67ca7072571.jpg</div>
  <div>students:[
    1,
    3
    ]</div>
  <div>available:true</div>
  <div>address:{
    "street": "1",
    "suburb": "Baulkham Hills",
    "state": "NSW",
    "postCode": 2153
    }</div>
</div>
`;
